### PR TITLE
[8.x] Allow queue:work to shut down the framework when exiting

### PR DESCRIPTION
**Pull request by fideloper**
## What

Following this discussion: [`app()->terminating()` not (always) called with `queue:work`](https://github.com/laravel/framework/discussions/32779), `queue:work` uses `exit($status)` when stopping the daemon. 

As a result, callbacks such as `app()->terminate()` are not called (and I assume other minor framework shutdown code, although given how long it's behaved like this, it may not be a big deal).

This PR allows the `Queue\Worker::daemon()` to return an exit code when breaking out of the `while (true)` loop, rather than exiting.

## Why

That means jobs that include code such as:

```php
app()->terminating(function() {
    Log::info("{$this->job->getJobId()} is running on app termination");
});
```

... will function when the queue is stopped.

## This is minor

This only matters when using `artisan queue:work` without `--once` (`queue:listen` doesn't require this).

Even then, it doesn't always make sense to register `app()->terminating(fn...)` callbacks in job context when you have no idea when a `queue:work` command will actually be stopped (and thus don't know when those callbacks will be used).

## Tests

One test was adjusted for this change, and one test was added to test out the memory limit exit code (`12`).

Testing `app()->terminate()` callbacks is tougher as the test environment `app()` is an instance of `Container` and doesn't have the `terminate()` callback methods.

Creating a new Laravel 8 app, applying this patch, does work. An example job:

```php
<?php

namespace App\Jobs;

use Illuminate\Support\Facades\Log;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;

class AddTerminatingThing implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /**
     * Create a new job instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        app()->terminating(function() {
            Log::info('we are terminating');
        });
        sleep(1);
    }
}
```

### CODE REVIEW

**Comment by fideloper** [comment]
I can fix the tests here pretty easily. Would love some feedback on if anything thinks this is a general waste of time as the use case seems minor.

I myself thought I wanted this but since I can't control when `queue:work` is stopped and thus don't know when shutdown callbacks would be called, I didn't end up going down this path.

**Comment by taylorotwell** [comment]
Seems generally OK to me.

File: src/Illuminate/Queue/Worker.php [hunk 1]
```diff
@@ -646,14 +658,14 @@ public function memoryExceeded($memoryLimit)
     /**
      * Stop listening and bail out of the script.
      *
-     * @param  int  $status
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Worker.php:649]: Please revert the whitespace changes here.

File: src/Illuminate/Queue/Worker.php [hunk 2]
```diff
@@ -123,7 +127,10 @@ public function daemon($connectionName, $queue, WorkerOptions $options)
             // if it is we will just pause this worker for a given amount of time and
             // make sure we do not need to kill this worker process off completely.
             if (! $this->daemonShouldRun($options, $connectionName, $queue)) {
-                $this->pauseWorker($options, $lastRestart);
+                $exitCode = $this->pauseWorker($options, $lastRestart);
+                if ($exitCode  !== false) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Worker.php:131]: One too many spaces. ;)

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
Looks good. Some very minor comments. ;)

**Comment by derekmd** [comment]
I wonder if this queue worker `$app->terminating()` gap is what causes https://github.com/laravel/telescope/issues/577 ? `ListensForStorageOpportunities::storeEntriesBeforeTermination()` registers Telescope entry updates that may be skipped if `exit()` is called prematurely. https://github.com/laravel/telescope/blob/ad74ce0499f5facb73f85ea0a0263b0d8cc0838c/src/ListensForStorageOpportunities.php#L43-L45 Although the docblock explicitly states:

> This handles storing entries for HTTP requests and Artisan commands.

I was never able to reproduce that Telescope job status bug but this pull request may be worth pursuing if it's concluded `exit()` was the cause.

**Comment by taylorotwell** [comment]
@derekmd if someone calls the PHP "exit" function then I'm not sure terminating callbacks would fire in that case.

**Pull request closed without merging**